databroker/experimental/server_ext.py

```diff
@@ -0,0 +1,501 @@
+
+
+@router.put("/array/full/{path:path}")
+async def put_array_full(
+    request: Request,
+    entry=Security(entry, scopes=["write:data", "write:metadata"]),
+):
+    if not isinstance(entry, WritingArrayAdapter):
+        raise HTTPException(
+            status_code=404, detail="This path cannot accept this array."
+        )
+    data = await request.body()
+    entry.put_data(data)
+    return json_or_msgpack(request, None)
+
+
+@router.put("/dataframe/full/{path:path}")
+async def put_dataframe_full(
+    request: Request,
+    entry=Security(entry, scopes=["write:data", "write:metadata"]),
+):
+    if not isinstance(entry, WritingDataFrameAdapter):
+        raise HTTPException(
+            status_code=404, detail="This path cannot accept this dataframe."
+        )
+    data = await request.body()
+    entry.put_data(data)
+    return json_or_msgpack(request, None)
+
+
+def array_raise_if_inactive(method):
+    def inner(self, *args, **kwargs):
+        if self.array_adapter is None:
+            raise ValueError("Not active")
+        else:
+            return method(self, *args, **kwargs)
+
+    return inner
+
+
+def dataframe_raise_if_inactive(method):
+    def inner(self, *args, **kwargs):
+        if self.dataframe_adapter is None:
+            raise ValueError("Not active")
+        else:
+            return method(self, *args, **kwargs)
+
+    return inner
+
+
+class WritingArrayAdapter:
+    structure_family = "array"
+
+    def __init__(self, collection, directory, doc):
+        self.collection = collection
+        self.directory = directory
+        self.doc = Document(**doc)
+        self.array_adapter = None
+        if self.doc.data_url is not None:
+            path = self.doc.data_url.path
+            if platform == "win32" and path[0] == "/":
+                path = path[1:]
+
+            file = h5py.File(path)
+            dataset = file["data"]
+            self.array_adapter = ArrayAdapter(dask.array.from_array(dataset))
+        elif self.doc.data_blob is not None:
+            self.array_adapter = ArrayAdapter(dask.array.from_array(self.doc.data_blob))
+
+    @property
+    def structure(self):
+        return ArrayStructure.from_json(self.doc.structure)
+
+    @property
+    def metadata(self):
+        return self.doc.metadata
+
+    @array_raise_if_inactive
+    def read(self, *args, **kwargs):
+        return self.array_adapter.read(*args, **kwargs)
+
+    @array_raise_if_inactive
+    def read_block(self, *args, **kwargs):
+        return self.array_adapter.read_block(*args, **kwargs)
+
+    def microstructure(self):
+        return self.array_adapter.microstructure()
+
+    def macrostructure(self):
+        return self.array_adapter.macrostructure()
+
+    def put_data(self, body):
+        # Organize files into subdirectories with the first two
+        # charcters of the uid to avoid one giant directory.
+        path = self.directory / self.doc.uid[:2] / self.doc.uid
```

Comment:
maybe add the `.hdf5` extension to the path so that the data store is more easily readable to external programs